packages/SwingSet/src/kernel/kernel.js

```diff
@@ -28,16 +28,124 @@ function abbreviateReviver(_, arg) {
 
 export default function buildKernel(kernelEndowments) {
   const {
-    setImmediate,
+    waitUntilQuiescent,
     hostStorage,
-    runEndOfCrank,
     vatAdminVatSetup,
     vatAdminDevSetup,
+    replaceGlobalMeter,
+    transformMetering,
+    transformTildot,
   } = kernelEndowments;
   insistStorageAPI(hostStorage);
   const { enhancedCrankBuffer, commitCrank } = wrapStorage(hostStorage);
   const kernelKeeper = makeKernelKeeper(enhancedCrankBuffer);
 
+  // It is important that tameMetering() was called by application startup,
+  // before install-ses. We expect the controller to run tameMetering() again
+  // (and rely upon its only-once behavior) to get the control facet
+  // (replaceGlobalMeter), and pass it to us through kernelEndowments.
+
+  // TODO: be more clever. Only refill meters that were touched. Use a
+  // WeakMap. Drop meters that vats forget. Minimize the fast path. Then wrap
+  // more (encapsulate importBundle?) and provide a single sensible thing to
+  // vats.
+  let complainedAboutGlobalMetering = false;
+  const allRefillers = new Set(); // refill() functions
+  function makeGetMeter(options = {}) {
+    const { refillEachCrank = true, refillIfExhausted = true } = options;
+
+    if (!replaceGlobalMeter && !complainedAboutGlobalMetering) {
+      console.error(
+        `note: makeGetMeter() cannot enable global metering, app must import @agoric/install-metering-and-ses`,
+      );
+      // to fix this, your application program must
+      // `import('@agoric/install-metering-and-ses')` instead of
+      // `import('@agoric/install-ses')`
+      complainedAboutGlobalMetering = true;
+    }
+
+    // zoe's importBundle(dapp-encouragement contract) causes babel to use
+    // about 5.4M computrons and 6.4M allocatrons (total 11.8M) in a single
+    // crank, so the default of 1e7 is insufficient
+    const FULL = 1e8;
+    const { meter, refillFacet } = makeMeter({
+      budgetAllocate: FULL,
+      budgetCompute: FULL,
+    });
+    function refill() {
+      // console.error(`-- METER REFILL`);
+      // console.error(`   allocate used: ${FULL - refillFacet.getAllocateBalance()}`);
+      // console.error(`   compute used : ${FULL - refillFacet.getComputeBalance()}`);
+      if (meter.isExhausted() && !refillIfExhausted) {
+        return;
+      }
+      refillFacet.allocate();
+      refillFacet.compute();
+      refillFacet.stack();
+    }
+    if (refillEachCrank) {
+      allRefillers.add(refill);
+    }
+    let meterUsed = false; // mostly for testing
+    function getMeter() {
```

Comment:
`getMeter` is called with an argument of `true` if we are supposed to return the meter but not install it in the globals.  `getMeter(true)` is used by the metering transform to bill the source string to the meter without replacing the global meter.

The failure to implement this mode is the reason your attempt to disable the global meter before transforming didn't work.

When we fix the metering transform not to use `getMeter` directly, this won't be a problem anymore.

Reply:
```suggestion
    function getMeter(dontReplaceGlobalMeter = false) {
```

Reply:
aha! Ok, I've applied that. Was that documented anywhere that I should have seen? I'll add a followup ticket to update the transform-metering README with more notes.

Reply:
Not documented, but present in the implementation of `getMeter` in `spawner` and `zoe`.